tls/s2n_server_hello.c

```diff
@@ -166,6 +166,10 @@ static int s2n_server_hello_parse(struct s2n_connection *conn)
         }
     }
 
+    if (s2n_hello_retry_validate(conn) == S2N_SUCCESS) {
```

Comment:
Instead of validating the retry message again, wouldn't it make more sense to check that this is a retry handshake? It might also be easier to follow this logic if you do it above where actual_protocol_version is set, to make it clear that's why this check is happening.

Additionally, maybe this check belongs in handshake_io when we call set_handshake_type?

Reply:
Yes, we could alternate this check to `s2n_is_hello_retry_handshake(conn)`, I would prefer to keep the check within this function and not delay the check.

Reply:
I wouldn't think of it as delaying the check. It's about scope of responsibility. set_handshake_type, not the server_hello, determines what kind of handshake we should be doing.

Making this check in server_hello is fine, but it would arguably make more sense in set_handshake_type. set_handshake_type is called by server_hello_recv, so it's not even much of a "delay".

Reply:
Alright I could move the check to https://github.com/awslabs/s2n/blob/master/tls/s2n_handshake_io.c#L532-L534. 

```
    if (conn->handshake.handshake_type & HELLO_RETRY_REQUEST) {
        ENSURE_POSIX(conn->actual_protocol_version >= S2N_TLS13, S2N_ERR_INVALID_HELLO_RETRY);
        conn->handshake.handshake_type = HELLO_RETRY_REQUEST;
    }
```